docstore/driver/codec.go

```diff
@@ -388,8 +388,10 @@ func decode(v reflect.Value, d Decoder) error {
 		if err != nil {
 			return err
 		}
-		v.Set(reflect.ValueOf(val))
-		return nil
+		if v.Type().AssignableTo(reflect.TypeOf(val)) {
+			v.Set(reflect.ValueOf(val))
+			return nil
+		}
```

Comment:
Is it OK to fall through if it's not assignable? I was kind of surprised when the tests passed. Maybe one of the options below ends up working?

For the new tests, this panics because `**not assignable uuid.UUID to []uint8**`. `uuid.UUID` is an array, not a slice.

Reply:
Here is what's happening, and why falling through works for mongo.

First, throughout this code, `v` is the destination and the value being decoded is inside `d`. We repeatedly ask `d` for the value in different forms. `AsSpecial` asks `d` to deal with any special cases, and `AsBytes` asks `d` for the value as a byte slice, if it can do that.

If `AsSpecial` returns true as its first return value (`done` here), then `val` is now the value to set into `v`, not whatever is in `d`. At that point, `val` must be used; falling through is not in general correct.
I thought about falling through but using `val` as the new value instead of what is inside `d`, but that is infeasible.

There is a mongo type called `primitive.Binary` which wraps a byte slice. The `AsSpecial` method unwraps it, but _so does `AsBytes`_. So in that particular case, falling through works because later, when we call `AsBytes`, we get the right value.

I think the surprising fix here is to remove the `primitive.Binary` case from the mongo driver's `AsSpecial`, and return an error if this assignability check fails. In general, if a driver can convert a decoder value using `AsBytes` or any similar method, it should not handle it in `AsSpecial`.
 
Let me know if that causes other problems.

Reply:
Thanks for the info. I tried that (removing the `primitive.Binary` case from the mongo `AsSpecial`, and returning an error here if `IsAssignable` return false).

The tests fail with:

```
ABOUT TO RETURN ERROR val 2019-03-26 17:00:00.005 -0700 PDT
v.Type is interface {}
TypeOf(val) is time.Time

decoding time.Time: cannot set type interface {} to 1553644800005 (code=InvalidArgument)
```

The code looks like this:

```
                if v.Type().AssignableTo(reflect.TypeOf(val)) {
                        v.Set(reflect.ValueOf(val))
                        return nil
                }
                fmt.Printf("ABOUT TO RETURN ERROR val %v\n", val)
                fmt.Printf("v.Type is %v\n", v.Type())
                fmt.Printf("TypeOf(val) is %v\n", reflect.TypeOf(val))
                return decodingError(v, d)
```

Interestingly, if I just revert this change entirely and always call `Set`, it seems to work -- so `Set` does something (or at least, doesn't panic) if the `AssignableTo` check returns false.....

Reply:
Oh, I think you have it reversed. You want to know if val is assignable to v.

Reply:
The order of the args to AssignableTo is the reverse of that of an assignment statement, hence the confusion.

Reply:
Arg. :-)
Fixed. Works now.